rpmlb/builder/base.py

```diff
@@ -126,31 +163,72 @@ def edit_spec_file(target_path: Path):
 
             yield source_file, target_file
 
-    def edit_spec_file_by_macros(self, spec_file, macros_dict):
-        if not isinstance(macros_dict, dict):
-            return ValueError('macros should be dict object.')
-
-        with self.edit_spec_file(spec_file) as (fh_r, fh_w):
-            for key in list(macros_dict.keys()):
-                value = macros_dict[key]
-                if value is None or str(value) == '':
-                    raise ValueError('macro is invalid in {0}.'.format(key))
-                content = '%global {0} {1}\n'.format(key, value)
-                fh_w.write(content)
-            fh_w.write('\n')
-            fh_w.write(fh_r.read())
-
-    def edit_spec_file_by_replaced_macros(self, spec_file, macros_dict):
-        if not isinstance(macros_dict, dict):
-            return ValueError('macros should be dict object.')
-
-        with self.edit_spec_file(spec_file) as (fh_r, fh_w):
-            for line in fh_r:
-                line = line.rstrip()
-                for key in list(macros_dict.keys()):
-                    value = macros_dict[key]
-                    pattern = r'^%global\s+{0}\s+[^\s]+$'.format(key)
-                    replaced_str = r'%global {0} {1}'.format(key, value)
-
-                    line = re.sub(pattern, replaced_str, line)
-                fh_w.write(line + '\n')
+            # Ensure that all unprocessed source contents
+            # are written to the target
+            target_file.write(source_file.read())
+
+    @staticmethod
+    def add_macros(source: Iterator[str], macros: Mapping[str, str]):
+        """Add macro definitions to the source stream.
+
+        Keyword arguments:
+            source: The source file iterator.
+            macros: Mapping of macro name to macro definition
+                    for all macros to add.
+
+        Yields:
+            Lines of the modified file.
+        """
+
+        # Prepend all definitions before the stream
+        for name, value in macros.items():
+            yield '%global {name} {value}\n'.format(name=name, value=value)
```

Comment:
It is. `yield` turns the function into a generator. `content_stream` is that generator, which I subsequently manipulate.